apps/server/src/modules/tldraw/service/tldraw.ws.service.ts

```diff
@@ -2,52 +2,81 @@ import { Injectable } from '@nestjs/common';
 import { ConfigService } from '@nestjs/config';
 import WebSocket from 'ws';
 import { applyAwarenessUpdate, encodeAwarenessUpdate, removeAwarenessStates } from 'y-protocols/awareness';
-import { encoding, decoding, map } from 'lib0';
+import { decoding, encoding, map } from 'lib0';
 import { readSyncMessage, writeSyncStep1, writeUpdate } from 'y-protocols/sync';
-import { Persitence, WSConnectionState, WSMessageType } from '../types';
+import { Buffer } from 'node:buffer';
+import { Redis } from 'ioredis';
+import { Logger } from '@src/core/logger';
+import { applyUpdate } from 'yjs';
+import {
+	RedisGeneralErrorLoggable,
+	RedisPublishErrorLoggable,
+	WebsocketCloseErrorLoggable,
+	WebsocketMessageErrorLoggable,
+	WsSharedDocErrorLoggable,
+} from '../loggable';
 import { TldrawConfig } from '../config';
-import { WsSharedDocDo } from '../domain/ws-shared-doc.do';
+import { AwarenessConnectionsUpdate, WSConnectionState, WSMessageType } from '../types';
+import { WsSharedDocDo } from '../domain';
 import { TldrawBoardRepo } from '../repo';
 import { MetricsService } from '../metrics';
 
 @Injectable()
 export class TldrawWsService {
 	public pingTimeout: number;
 
-	public persistence: Persitence | null = null;
+	private gcEnabled: boolean;
 
 	public docs = new Map();
 
+	public readonly sub: Redis;
+
+	private readonly pub: Redis;
+
 	constructor(
 		private readonly configService: ConfigService<TldrawConfig, true>,
 		private readonly tldrawBoardRepo: TldrawBoardRepo,
+		private readonly logger: Logger,
 		private readonly metricsService: MetricsService
 	) {
+		this.logger.setContext(TldrawWsService.name);
 		this.pingTimeout = this.configService.get<number>('TLDRAW_PING_TIMEOUT');
-	}
+		this.gcEnabled = this.configService.get<boolean>('TLDRAW_GC_ENABLED');
+		const redisUri: string = this.configService.get<string>('REDIS_URI');
+
+		if (!redisUri) {
+			throw new Error('REDIS_URI is not set');
+		}
+
+		this.sub = new Redis(redisUri, {
+			maxRetriesPerRequest: null,
+		});
+		this.sub.on('error', (err) => this.logger.warning(new RedisGeneralErrorLoggable('SUB', err)));
 
-	public setPersistence(persistence_: Persitence): void {
-		this.persistence = persistence_;
+		this.pub = new Redis(redisUri, {
+			maxRetriesPerRequest: null,
+		});
+		this.pub.on('error', (err) => this.logger.warning(new RedisGeneralErrorLoggable('PUB', err)));
 	}
 
 	/**
 	 * @param {WsSharedDocDo} doc
 	 * @param {WebSocket} ws
 	 */
 	public closeConn(doc: WsSharedDocDo, ws: WebSocket): void {
-		if (doc.conns.has(ws)) {
-			const controlledIds = doc.conns.get(ws) as Set<number>;
-			doc.conns.delete(ws);
+		if (doc.connections.has(ws)) {
```

Comment:
I am unsure if the code work like you expect. 
The ws.close(); should always executed before any code this.tldrawBoardRepo can be arrived the first .then

Reply:
that's correct, operations in .then do not do any work on the websocket so it works as expected